paimon-core/src/main/java/org/apache/paimon/jdbc/JdbcCatalogLock.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.jdbc;
+
+import org.apache.paimon.catalog.CatalogLock;
+import org.apache.paimon.utils.TimeUtils;
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+import static org.apache.paimon.options.CatalogOptions.LOCK_ACQUIRE_TIMEOUT;
+import static org.apache.paimon.options.CatalogOptions.LOCK_CHECK_MAX_SLEEP;
+
+/** Jdbc catalog lock. */
+public class JdbcCatalogLock implements CatalogLock {
+    private final JdbcClientPool connections;
+    private final long checkMaxSleep;
+    private final long acquireTimeout;
+    private final String catalogName;
+
+    public JdbcCatalogLock(
+            JdbcClientPool connections,
+            String catalogName,
+            long checkMaxSleep,
+            long acquireTimeout) {
+        this.connections = connections;
+        this.checkMaxSleep = checkMaxSleep;
+        this.acquireTimeout = acquireTimeout;
+        this.catalogName = catalogName;
+    }
+
+    @Override
+    public <T> T runWithLock(String database, String table, Callable<T> callable) throws Exception {
+        String lockUniqueName = String.format("%s.%s.%s", catalogName, database, table);
+        lock(lockUniqueName);
+        try {
+            return callable.call();
```

Comment:
What happens if the lock expires before the commit is completed? Will there be conflicts? Does this break the atomicity of commits?

For example, let's consider the first column as logical timestamp, and the second column is the events that occurred at that time:
1   client-1 successfully aquired a lock of table `a`
2   client-1 starts committing but gets stuck for some reason
5   the lock of table `a ` timeouts
6   client-2 cleaned the expired lock of table `a` and acquired a new lock of table `a`
7   client-2 starts committing
8   client-2 made a successful committment
9   client-1 can start committing

In this case, client-1's commit should fail, but here client-1 will continue to commit. If the underlying file system allows overwriting of files with the same name, then client-1's commit will overwrite client-2's commit.

Reply:
> What happens if the lock expires before the commit is completed? Will there be conflicts? Does this break the atomicity of commits?
> 
> For example, let's consider the first column as logical timestamp, and the second column is the events that occurred at that time: 1 client-1 successfully aquired a lock of table `a` 2 client-1 starts committing but gets stuck for some reason 5 the lock of table `a ` timeouts 6 client-2 cleaned the expired lock of table `a` and acquired a new lock of table `a` 7 client-2 starts committing 8 client-2 made a successful committment 9 client-1 can start committing
> 
> In this case, client-1's commit should fail, but here client-1 will continue to commit. If the underlying file system allows overwriting of files with the same name, then client-1's commit will overwrite client-2's commit.

I understand that this should require checking when writing to ensure the order of submissions, which should not be related to distributed locks, or setting a reasonable timeout time

Reply:
You can start a new thread renewal timeout to solve the problem, but you can also set a reasonable timeout to solve it

Reply:
For the current strategy, it is difficult for this situation to occur because currently using lock is only renaming the file, and this operation should be completed quickly.